PEP 691: Draft PEP for Simple JSON API (#2578)
python/peps: pep-0691.rst

PEP: 691
Title: JSON-based Simple API for Python Package Indexes
Author: Donald Stufft <donald@stufft.io>,
        Pradyun Gedam <pradyunsg@gmail.com>,
        Cooper Lees <me@cooperlees.com>,
        Dustin Ingram <di@python.org>
Status: Draft

> Type: Informational

Reviewer: Are you sure it should be Informational? It reads like a standards-track PEP.

Author: This is based on https://peps.python.org/pep-0503/ being informational, but I'm fine to change it if this doesn't make sense anymore.

Reviewer: Interesting, I feel like it should not be Informational because you are proposing a new standard that multiple tools will have to implement. An Informational PEP would be appropriate if you're just documenting how this API works in Warehouse.

But ultimately that's a decision for the packaging community, so I won't hold up merging the PEP over it. I can merge it now if you like, or wait for your co-authors or others to chime in.

Reviewer 2: PEP 503 was written post the Simple HTML already existing so I second @JelleZijlstra here - We should probably change.

Author: I agree, but let's see what the wider community thinks in discussion. Fine to merge now!

Author: ```suggestion
Type: Standards Track
```

Reviewer 3: There's like a 95% chance I picked informational on PEP 503 because I didn't care very much and that seemed fine and nobody called me out on it.

As an example::

    {
      "name": "holygrail",
      "files": [
        {
          "filename": "holygrail-1.0.tar.gz",
          "url": "https://example.com/files/holygrail-1.0.tar.gz",
          "hashes": {"sha256": "...", "blake2b": "..."},
          "requires-python": ">=3.7",
          "yanked": "Had a vulnerability"
        },
        {
          "filename": "holygrail-1.0-py3-none-any.whl",
          "url": "https://example.com/files/holygrail-1.0-py3-none-any.whl",
          "hashes": {"sha256": "...", "blake2b": "..."},
          "requires-python": ">=3.7",
          "dist-info-metadata-available": true
        },
      ]
    }

In addition to the above, the following constraints are placed on the API:

* While JSON doesn't natively support an URL type, any value that represents an
  URL in this API may be either absolute or relative as long as they point to
  the correct location. If relative, they are relative to the current URL as if
  it were HTML.

* Additional keys may be added to any dictionary objects in the API responses
  and clients **MUST** ignore keys that they don't understand.

* By default, any hash algorithm available via `hashlib

>   <https://docs.python.org/3/library/hashlib.html>` (specifically any that can

Reviewer: ```suggestion
  <https://docs.python.org/3/library/hashlib.html>`_ (specifically any that can
```

  determine if the JSON URL exists or not.

  The most naive implementation of this would be to request the JSON URL and fall back
  to the HTML URL for *every* single request, but that would be horribly performant
  and violate the goal of minimal additional HTTP requests.

  The most likely implementation of this would be to make some sort of repository level
  configuration file that somehow indicates what is supported. We would have the same
  namespace problem as above, with the same solution, something like ``/_config.json``
  or so could hold that data, and a client could first make an HTTP request to that,
  and if it exists pull it down and parse it to learn about the capabilities of this
  particular repository.
- The use of ``Accept`` also allows us to add versioning into this field

All being said, it is the opinion of this PEP that those three issues combined make
using separate API routes a less desirable solution than relying on content
negotiation to select the most ideal representation of the data.

Appendix 1: Survey of use cases to cover
========================================

This was done through a discussion between ``pip`` and ``bandersnarch``

> maintainers, where the two first potential use cases for the new API. This is

Reviewer: Some missing word here

Author: ```suggestion
maintainers, who were the two first potential user for the new API. This is
```

how they use the Simple + JSON APIs today:

- ``pip``:

  - List of all files for a particular release
  - Metadata of each individual artifact:

    - was it yanked? (`data-yanked`)
    - what's the python-requires? (`data-python-requires`)
    - what's the hash of this file? (currently, hash in URL)
    - Full metadata (`data-dist-info-metadata`)
    - [Bonus] what are the declared dependencies, if available (list-of-strings, null if unavailable)?

- ``bandersnatch`` - Only uses legacy JSON API + XMLRPC today:

  - Generates Simple HTML rather than copying from PyPI

    - Maybe this changes with the new API and we verbatim pull these API assets from PyPI

  - List of all files for a particular release.

    - Workout URL for release files to download

  - Metadata of each individual artifact.

    - Write out the JSON to mirror storage today (disk/S3)

>         - Required metadata used (via Package class - https://github.com/pypa/bandersnatch/blob/main/src/bandersnatch/package.py):

Reviewer: This renders a bit weirdly: https://pep-previews--2578.org.readthedocs.build/pep-0691/#appendix-1-survey-of-use-cases-to-cover

Author: I'm not sure what's causing this. Is this a limitation of rST, an expected style for PEPs, or something syntactically wrong?

Reviewer: I think it's indented by one space too many.

Reviewer: I applied the change, will confirm it looks OK in the preview and then merge.

>         - Required metadata used (via Package class - https://github.com/pypa/bandersnatch/blob/main/src/bandersnatch/package.py):
> 
>           - metadata["info"]
>           - metadata["last_serial"]
>           - metadata["releases"]
> 
>             - digests
>             - URL

Reviewer: ```suggestion
      - Required metadata used (via Package class - https://github.com/pypa/bandersnatch/blob/main/src/bandersnatch/package.py):

        - metadata["info"]
        - metadata["last_serial"]
        - metadata["releases"]

          - digests
            - URL
```
